File: src/main/java/org/stt/time/Interval.java
package org.stt.time;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Objects;

public class Interval {
    private final LocalDateTime start;
    private final LocalDateTime end;

    private Interval(LocalDateTime start, LocalDateTime end) {
        Objects.requireNonNull(start);
        Objects.requireNonNull(end);
        this.start = start;
        this.end = end;
    }

    public LocalDateTime getStart() {
        return start;
    }

    public LocalDateTime getEnd() {
        return end;
    }

    public static Interval between(LocalDateTime start, LocalDateTime end) {
        return new Interval(start, end);
    }

    public static Interval ofDay(LocalDate day) {
        return new Interval(day.atStartOfDay(), day.plusDays(1).atStartOfDay());
    }

    public static Interval between(LocalDate start, LocalDate end) {
        return new Interval(start.atStartOfDay(), end.plusDays(1).atStartOfDay());
    }

    public Interval withEnd(LocalDateTime newEnd) {
        return new Interval(start, newEnd);
    }
}

File: src/main/kotlin/org/stt/time/Interval.kt
package org.stt.time

import java.time.LocalDate
import java.time.LocalDateTime

class Interval internal constructor(val start: LocalDateTime, val end: LocalDateTime) {
    fun withEnd(newEnd: LocalDateTime): Interval {
        return Interval(start, newEnd)
    }
}


fun LocalDate.asInterval() = Interval(atStartOfDay(), plusDays(1).atStartOfDay())
infix fun LocalDate.until(end: LocalDate) = Interval(atStartOfDay(), end.plusDays(1).atStartOfDay())
infix fun LocalDateTime.until(end: LocalDateTime) = Interval(this, end)